protocol/types.go

```diff
@@ -187,3 +190,174 @@ type VersionResponse struct {
 	// Build contains the timestamp at the time of the build.
 	Build time.Time `json:"build"`
 }
+
+// Node is a node in a UAST.
+//
+//proteus:generate
+type Node struct {
+	// InternalType is the internal type of the node in the AST, in the source
+	// language.
+	InternalType string `json:",omitempty"`
+	// Properties are arbitrary, language-dependent, metadata of the
+	// original AST.
+	Properties map[string]string `json:",omitempty"`
+	// Children are the children nodes of this node.
+	Children []*Node `json:",omitempty"`
+	// Token is the token content if this node represents a token from the
+	// original source file. If it is empty, there is no token attached.
+	Token string `json:",omitempty"`
+	// StartPosition is the position where this node starts in the original
+	// source code file.
+	StartPosition *uast.Position `json:",omitempty"`
+	// EndPosition is the position where this node ends in the original
+	// source code file.
+	EndPosition *uast.Position `json:",omitempty"`
+	// Roles is a list of Role that this node has. It is a language-independent
+	// annotation.
+	Roles []role.Role `json:",omitempty"`
+}
+
+// NewNode creates a new empty *Node.
+func NewNode() *Node {
+	return &Node{
+		Properties: make(map[string]string, 0),
+		Roles:      []role.Role{role.Unannotated},
+	}
+}
+
+// String converts the *Node to a string using pretty printing.
+func (n *Node) String() string {
+	buf := bytes.NewBuffer(nil)
+	err := Pretty(n, buf, IncludeAll)
+	if err != nil {
+		return "error"
+	}
+
+	return buf.String()
+}
+
+const (
+	// InternalRoleKey is a key string uses in properties to use the internal
+	// role of a node in the AST, if any.
+	InternalRoleKey = "internalRole"
+)
+
+// ToNode converts a generic AST node to Node object used in the protocol.
+func ToNode(n uast.Node) (*Node, error) {
+	nd, err := asNode(n, "")
+	if err != nil {
+		return nil, err
+	}
+	switch len(nd) {
+	case 0:
+		return nil, nil
+	case 1:
+		return nd[0], nil
+	default:
+		return &Node{Children: nd}, nil
+	}
+}
+
+func asNode(n uast.Node, field string) ([]*Node, error) {
```

Comment:
can you break this function in several ones, something like `arrayAsNode`, etc and call it from this function?